File: mobile/src/main/java/com/mattrubacky/monet2/splatnet/SplatnetRequest.java
package com.mattrubacky.monet2.splatnet;

import android.content.Context;
import android.os.Bundle;
import android.os.Parcelable;

import java.io.IOException;
import java.net.MalformedURLException;

import retrofit2.Call;
import retrofit2.Response;

/**
 * Created by mattr on 12/6/2017.
 */

public abstract class SplatnetRequest {
    protected Call call;

    protected abstract void manageResponse(Response response) throws IOException, SplatnetUnauthorizedException;
    public abstract void setup(Splatnet splatnet, String cookie, String uniqueID);

    public void run() throws SplatnetUnauthorizedException,MalformedURLException,IOException{
        if(shouldUpdate()){
            Response response = call.execute();
            if (response.isSuccessful()) {
                manageResponse(response);
            } else {
                throw new SplatnetUnauthorizedException("User Authentication Failed");
            }
        }else{
            handleNoUpdate();
        }
    }

    public Bundle result(Bundle bundle){
        return bundle;
    }

    public boolean shouldUpdate(){
        return true;
    }
    public void handleNoUpdate(){}
}

File: mobile/src/main/java/com/mattrubacky/monet2/backend/api/splatnet/SplatnetRequest.kt
package com.mattrubacky.monet2.backend.api.splatnet

import com.mattrubacky.monet2.backend.api.Request
import retrofit2.Response

abstract class SplatnetRequest<T>(splatnet:Splatnet,cookie:String,uniqueId:String): Request<T>() {
    override fun manageError(response: Response<T>) {
        error = if(response.code()==200){
            "Splatnet is down for Maintenance."
        }else{
            "User Authentication failed"
        }
    }
}